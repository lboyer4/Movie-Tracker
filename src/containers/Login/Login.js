import React, { Component } from 'react';
import { fetchUsers } from '../../utils/fetchUsers';
import { connect } from 'react-redux';
import { updateLogin } from '../../actions';
import { Redirect } from 'react-router-dom';
import { setFavorites } from '../../actions';
import PropTypes from 'prop-types';


export class Login extends Component {
	constructor() {
		super()
		this.state = {
			email: '',
			password: '',
			loggedIn: false,
			error: ''
		}
	}

	handleChange = (e) => {
		const { name, value } = e.target;
		this.setState({ [name] : value });
	};

	handleSubmit = (e) => {
		e.preventDefault();
		this.postLogin(this.state);
		this.setState({ email: '', password: '' });
	};

	postLogin = (state) => {
		const url = 'http://localhost:3000/api/users';
		const { email, password } = state;
		const options = {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json'
			},
			body: JSON.stringify({
				email, 
				password
			})
		};

		fetchUsers(url, options)
		.then(results => {
			if(results === 'error') {
				this.setState({error: '*Email and password do not match, please try again'})
			} else {
				this.setState({ loggedIn: true })
				this.props.updateLogin(
					results.data
          )
			}
		});
	};

	render() {
		if(this.state.loggedIn) {
			return ( 
				<Redirect to="/" />
				)
		}
		return (
			<form className="signin-form" onSubmit={this.handleSubmit}>
			<div className="create-header">
				<h1 className="form-title"> Log in </h1>
			</div>
				<h3 className="error-msg" >{this.state.error}</h3>
				<input 
					className="email-input"
					type="text"
					name="email"
					placeholder="email"
					value={this.state.email}
					onChange={this.handleChange}
				/>
				<input
					type="password"
					name="password"
					placeholder="password"
					value={this.state.password} 
					onChange={this.handleChange}
				/>
				<button className="submit-btn"> Submit </button>
			</form>
			);
	};
};

export const mapStateToProps = (state) => ({
	loggedIn: state.loggedIn
});

export const mapDispatchToProps = (dispatch) => ({
	updateLogin: (user) => dispatch(updateLogin(user)),
	setFavorites: (favorites) => dispatch(setFavorites(favorites))
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);

Login.propType = {
	history: PropTypes.object,
	location: PropTypes.object,
	loggedIn: PropTypes.object,
	match:PropTypes.object,
	setFavorites: PropTypes.func,
	updateLogin: PropTypes.func,
}