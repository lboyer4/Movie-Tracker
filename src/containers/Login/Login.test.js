import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { setFavorites, updateLogin } from '../../actions'


describe('Login', () => {
	it('should match snapshot', () => {
		const wrapper = shallow(
			<Login />
		)
		expect(wrapper).toMatchSnapshot();
	});

	describe('handleChange', () => {
		it('should set state when handle change is called', () => {
			const mockUserEvent = {
				target: {
					name: 'email',
					value: 'puppies@gmail'
				}
			}

			const wrapper = shallow(
			<Login />
			)
			wrapper.instance().handleChange(mockUserEvent)
			expect(wrapper.state().email).toEqual('puppies@gmail')
		});
	});

	describe('handleSubmit', () => {
		it.skip('should call post login', () => {
			const wrapper = shallow(
			<Login />
			)
			const mockSubmit = {
				preventDefault: () => {}
			};
			const mockState = {
				email: 'puppies@gmail'
			}
			wrapper.setState(mockState);
			wrapper.instance().handleSubmit(mockSubmit);
			expect(wrapper.state().email).toEqual(mockState)
		})
	})
	describe('mapStateToProps', () => {

		it('should return a user object and a favorites array', () => {

			const mockState = {
				loggedIn: {user: 'name'},
			}

			const expected = {
				loggedIn: mockState.loggedIn,
			}

			const mappedProps = mapStateToProps(mockState)

			expect(mappedProps).toEqual(expected)
		})
	})

	describe('mapDispatchToProps', () => {

		it('should call dispatch when using a function from MDTP', () => {

			const mockDispatch = jest.fn();

			const actionToDispatch = updateLogin({name: 'joe'})

			const mappedProps= mapDispatchToProps(mockDispatch)

			mappedProps.updateLogin({name: 'joe'})

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
		})

	it('should call dispatch when using a function from MDTP', () => {
					const mockDispatch = jest.fn();

			const actionToDispatch = setFavorites(['favorite'])

			const mappedProps= mapDispatchToProps(mockDispatch)

			mappedProps.setFavorites(['favorite'])

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
	})
	})
})