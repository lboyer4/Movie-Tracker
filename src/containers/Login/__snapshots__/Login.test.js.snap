// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<form
  className="signin-form"
  onSubmit={[Function]}
>
  <div
    className="create-header"
  >
    <h1
      className="form-title"
    >
       Log in 
    </h1>
  </div>
  <h3
    className="error-msg"
  />
  <input
    className="email-input"
    name="email"
    onChange={[Function]}
    placeholder="email"
    type="text"
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    placeholder="password"
    type="password"
    value=""
  />
  <button
    className="submit-btn"
  >
     Submit 
  </button>
</form>
`;
